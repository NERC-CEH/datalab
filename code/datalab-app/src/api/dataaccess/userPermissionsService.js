import axios from 'axios';
import logger from 'winston';
import { get } from 'lodash';
import config from '../config';
import axiosErrorHandler from '../util/errorHandlers';

const authPermissionsUrl = `${config.get('authorisationService')}/permissions`;
const authServiceStub = config.get('authorisationServiceStub');

function getUserPermissions(authZeroToken) {
  logger.debug('Requesting user permissions from auth service');

  if (authServiceStub) {
    return Promise.resolve([
      'project:stacks:create',
      'project:stacks:delete',
      'project:stacks:list',
      'project:stacks:open',
      'project:storage:create',
      'project:storage:delete',
      'project:storage:list',
      'project:storage:open',
      'project:users:list',
      'project:users:grant',
      'project:users:forbid',
    ]);
  }

  return axios.get(authPermissionsUrl, { headers: { authorization: authZeroToken } })
    .then(response => get(response, 'data.permissions') || [])
    .catch(axiosErrorHandler('Unable to get user permissions'));
}

export default getUserPermissions;
