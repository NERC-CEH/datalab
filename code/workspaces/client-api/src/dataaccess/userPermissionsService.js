import axios from 'axios';
import logger from 'winston';
import { get } from 'lodash';
import config from '../config';
import { axiosErrorHandler } from '../util/errorHandlers';

const authPermissionsUrl = `${config.get('authorisationService')}/permissions`;
const authServiceStub = config.get('authorisationServiceStub');

const stubPermissions = [
  'project:stacks:create',
  'project:stacks:delete',
  'project:stacks:list',
  'project:stacks:open',
  'project:stacks:edit',
  'project:storage:create',
  'project:storage:delete',
  'project:storage:list',
  'project:storage:open',
  'project:users:list',
];

function getUserPermissions(identity, token) {
  const userName = identity ? identity.userName : null;
  logger.debug(`Requesting user permissions from auth service for userName ${userName}`);

  if (authServiceStub) {
    return Promise.resolve(stubPermissions);
  }

  return axios.get(authPermissionsUrl, {
    headers: { authorization: token },
    params: { userName },
  })
    .then(response => get(response, 'data.permissions') || [])
    .catch(axiosErrorHandler('Unable to get user permissions'));
}

export default { getUserPermissions };
