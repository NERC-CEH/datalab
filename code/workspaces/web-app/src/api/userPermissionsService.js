import { gqlQuery } from './graphqlClient';
import errorHandler from './graphqlErrorHandler';

function getUserPermissions() {
  const query = `
    GetPermissions {
      userPermissions
    }`;

  return gqlQuery(query)
    .then(errorHandler('data.userPermissions'));
}

async function getOtherUserRoles(userId) {
  const query = `
    GetOtherUserRoles($userId: String!) {
      otherUserRoles(userId: $userId) {
        instanceAdmin,
        projectAdmin,
        projectUser,
        projectViewer,
        siteOwner { projectKey, name},
        notebookOwner { projectKey, name},
        storageAccess { projectKey, name}
      }
    }`;

  const otherUserRoles = await gqlQuery(query, { userId })
    .then(errorHandler('data.otherUserRoles'));

  return {
    userId,
    otherUserRoles,
  };
}

export default { getUserPermissions, getOtherUserRoles };
