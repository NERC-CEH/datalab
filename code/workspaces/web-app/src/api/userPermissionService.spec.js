import mockClient from './graphqlClient';
import userPermissionsService from './userPermissionsService';

jest.mock('./graphqlClient');

describe('userPermissionService', () => {
  beforeEach(() => {
    mockClient.clearResult();
  });

  it('userPermissionsService should build correct query and unpack the result', () => {
    mockClient.prepareSuccess({ userPermissions: 'expectedValue' });

    return userPermissionsService.getUserPermissions().then((response) => {
      expect(response).toEqual('expectedValue');
      expect(mockClient.lastQuery()).toMatchSnapshot();
    });
  });

  it('userPermissionsService should throw an error if the query fails', () => {
    mockClient.prepareFailure('error');

    return userPermissionsService.getUserPermissions().catch((error) => {
      expect(error).toEqual({ error: 'error' });
    });
  });
});
