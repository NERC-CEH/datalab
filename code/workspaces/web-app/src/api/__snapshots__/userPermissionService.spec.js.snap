// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userPermissionService userPermissionsService should build correct query and unpack the result 1`] = `
"
    GetPermissions {
      userPermissions
    }"
`;
